proxy/forward.go

```diff
@@ -0,0 +1,141 @@
+package proxy
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+)
+
+// ForwardConfig is the configuration to run a forward proxy
+type ForwardConfig struct {
+	CaCertPath string
+	CertPath string
+	KeyPath string
+	ListenPort int
+	MTLS bool
+}
+
+// ForwardEnabled checks the environmen to see if we're configured for Forward Proxy serving
+func ForwardEnabled() (*ForwardConfig, error) {
+	var exists bool
+	var err error
+	if os.Getenv("HABERDASHER_FWDPROXY") != "" {
+		config := ForwardConfig{}
+		if config.CaCertPath, exists = os.LookupEnv("HABERDASHER_TLS_CACERT"); !exists {
+			return nil, errors.New("HABERDASHER_TLS_CACERT not set")
+		}
+		if os.Getenv("HABERDASHER_FWDPROXY_MTLS") != "" {
+			if config.CertPath, exists = os.LookupEnv("HABERDASHER_TLS_CERT"); !exists {
+				return nil, errors.New("HABERDASHER_TLS_CERT not set")
+			}
+			if config.KeyPath, exists = os.LookupEnv("HABERDASHER_TLS_KEY"); !exists {
+				return nil, errors.New("HABERDASHER_TLS_KEY not set")
+			}
+		}
+		if listenPort, exists := os.LookupEnv("HABERDASHER_FWDPROXY_PORT"); !exists {
+			return nil, errors.New("HABERDASHER_FWDPROXY_PORT not set")
+		} else {
+			config.ListenPort, err = strconv.Atoi(listenPort)
+			if err != nil {
+				return nil, err
+			}
+		}
+		return &config, nil
+	}
+	return nil, nil
+}
+
+func forwardHandler(client *http.Client, writer http.ResponseWriter, req *http.Request) {
+	url := req.URL
+	// TODO: We need to intelligently copy HTTP headers onto the subrequest
+	proxyReq, err := http.NewRequest(req.Method, url.String(), req.Body)
+	// Do an SRV lookup on the service to see if it's secure or not
+	cname, addrs, err := net.LookupSRV("", "", url.Hostname())
```

Comment:
What should this do if there is an error here?

Reply:
Good question. My thought was that if we can't confirm that the request should be upgraded, we pass it along unmodified. In an ecosystem where the insecure port is not on the service, it would fail - in an ecosystem where both ports are on the service, it would continue unencrypted. Also, not all outbound requests will be to a `.svc.cluster.local` hostname.